chainerrl/experiments/evaluator.py

```diff
@@ -122,39 +152,68 @@ def batch_run_evaluation_episodes(
 
         # Agent observes the consequences
         agent.batch_observe(obss, rs, dones, resets)
-
         # Make mask. 0 if done/reset, 1 if pass
         end = np.logical_or(resets, dones)
         not_end = np.logical_not(end)
 
-        episode_returns.extend(episode_r[end])
-        episode_lengths.extend(episode_len[end])
-
-        if len(episode_returns) >= n_runs:
-            break
+        for index in range(len(end)):
+            if end[index]:
+                episode_returns[episode_indices[index]] = episode_r[index]
+                episode_lengths[episode_indices[index]] = episode_len[index]
+                episode_indices[index] = episode_idx
+                episode_idx += 1
 
         episode_r[end] = 0
         episode_len[end] = 0
-        obss = env.reset(not_end)
-
-    episode_returns = episode_returns[:n_runs]
-    episode_lengths = episode_lengths[:n_runs]
+        eval_episode_returns = []
+        eval_episode_lens = []
+        completed_episode = 0
+        total_time = 0
+        while completed_episode in episode_returns:
+            eval_episode_returns.append(episode_returns[completed_episode])
+            eval_episode_lens.append(episode_lengths[completed_episode])
+            total_time += episode_lengths[completed_episode]
+            completed_episode += 1
+            if n_steps is not None:
+                if total_time > n_steps:
+                    termination_conditions = True
+                    eval_episode_returns = eval_episode_returns[:-1]
+                    eval_episode_lens = eval_episode_lens[:-1]
+                    break
+            else:
+                if completed_episode >= n_episodes:
+                    termination_conditions = True
+                    eval_episode_returns = eval_episode_returns[:n_episodes]
+                    eval_episode_lens = eval_episode_lens[:n_episodes]
+                    break
+        # special case
+        if (n_steps is not None
+                and completed_episode == 0
+                and episode_len[0] >= n_steps):
+            eval_episode_returns = [episode_r[0]]
+            eval_episode_lens = [n_steps]
+            termination_conditions = True
+
+        if termination_conditions:
+            break
+        else:
+            obss = env.reset(not_end)
```

Comment:
The logic of making `eval_episode_returns` and `eval_episode_lens` looks wrong to me. To be unbiased, we need to use episodes with smaller `episode_idx` first, even when episodes with larger `episode_idx` finish earlier. That is,

- if n_steps is not None
  - check episode_idx=0,1,... sequentially until you first encounter an unfinished episode, then sum up their lengths (including last unfinished one). if >=n_steps, then terminate.
- if n_episodes is not None
  - check if episode_idx=0,1,...,(n_episodes-1) are all finished. if yes, then terminate. episode_idx=n_episodes,... should not be used at all.

Reply:
As discussed offline, the code is correct, though suboptimal. Will make the changes to improve code readability and efficiency.